src/graph/immutable_graph.cc

```diff
@@ -4,150 +4,179 @@
-  CHECK_EQ(indptr_data[0], 0);
-  CHECK_EQ(indptr_data[num_vertices], num_edges);
-  indptr.insert_back(indptr_data, num_vertices + 1);
-  indices.insert_back(indices_data, num_edges);
-  edge_ids.insert_back(edge_id_data, num_edges);
+CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids)
+  : indptr_(indptr), indices_(indices), edge_ids_(edge_ids) {
+  CHECK(IsValidIdArray(indptr));
+  CHECK(IsValidIdArray(indices));
+  CHECK(IsValidIdArray(edge_ids));
+  CHECK_EQ(indices->shape[0], edge_ids->shape[0]);
 }
 
-ImmutableGraph::CSR::CSR(IdArray indptr_arr, IdArray index_arr, IdArray edge_id_arr,
-                         const std::string &shared_mem_name) {
+CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids,
+         const std::string &shared_mem_name) {
 #ifndef _WIN32
-  size_t num_vertices = indptr_arr->shape[0] - 1;
-  size_t num_edges = index_arr->shape[0];
-  CHECK_EQ(num_edges, edge_id_arr->shape[0]);
-  size_t file_size = (num_vertices + 1) * sizeof(int64_t) + num_edges * sizeof(dgl_id_t) * 2;
-
-  auto mem = std::make_shared<runtime::SharedMemory>(shared_mem_name);
-  auto ptr = mem->create_new(file_size);
-
-  int64_t *addr1 = static_cast<int64_t *>(ptr);
-  indptr.init(addr1, num_vertices + 1);
-  void *addr = addr1 + num_vertices + 1;
-  dgl_id_t *addr2 = static_cast<dgl_id_t *>(addr);
-  indices.init(addr2, num_edges);
-  addr = addr2 + num_edges;
-  dgl_id_t *addr3 = static_cast<dgl_id_t *>(addr);
-  edge_ids.init(addr3, num_edges);
-
-  const int64_t *indptr_data = static_cast<int64_t*>(indptr_arr->data);
-  const dgl_id_t *indices_data = static_cast<dgl_id_t*>(index_arr->data);
-  const dgl_id_t *edge_id_data = static_cast<dgl_id_t*>(edge_id_arr->data);
-  CHECK_EQ(indptr_data[0], 0);
-  CHECK_EQ(indptr_data[num_vertices], num_edges);
-  indptr.insert_back(indptr_data, num_vertices + 1);
-  indices.insert_back(indices_data, num_edges);
-  edge_ids.insert_back(edge_id_data, num_edges);
-  this->mem = mem;
+  CHECK(IsValidIdArray(indptr));
+  CHECK(IsValidIdArray(indices));
+  CHECK(IsValidIdArray(edge_ids));
+  CHECK_EQ(indices->shape[0], edge_ids->shape[0]);
+  const int64_t num_verts = indptr->shape[0] - 1;
+  const int64_t num_edges = indices->shape[0];
+  const int64_t file_size = (num_verts + 1 + num_edges * 2) * sizeof(dgl_id_t);
+
+  IdArray sm_array = IdArray::EmptyShared(
+      shared_mem_name, {file_size}, DLDataType{kDLInt, 8, 1}, DLContext{kDLCPU, 0}, true);
+  // Create views from the shared memory array. Note that we don't need to save
+  //   the sm_array because the refcount is maintained by the view arrays.
+  indptr_ = sm_array.CreateView({num_verts + 1}, DLDataType{kDLInt, 64, 1});
+  indices_ = sm_array.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
+      (num_verts + 1) * sizeof(dgl_id_t));
+  edge_ids_ = sm_array.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
+      (num_verts + 1 + num_edges) * sizeof(dgl_id_t));
```

Comment:
now all arrays have a reference to the shared-memory object?

Reply:
In `NDArray::CreateView`, it increases the reference count of the base array (the `sm_array` in this context):
https://github.com/dmlc/dgl/blob/a7d0b8a0a416c0a253215e714ce274bb9dd88518/src/runtime/ndarray.cc#L136-L154

As a result, `sm_array` will not be freed untill all indices_, indptr_ and edge_ids_ are freed. Since `sm_array` holds a shared pointer to the shared memory object. That means we don't need to have that explicitly anymore.